File: compiler/nodes/expressions/number.go
package expressions

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/nodes/types"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type NumberExpressionNode struct {
	Value string
	Facts *database.Facts
}

func (node *NumberExpressionNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseNumberExpression(parser *syntax.Parser) (*NumberExpressionNode, *syntax.Error) {
	span := parser.Spanned()

	value, err := parser.Token("Number")
	if err != nil {
		return nil, err
	}

	return &NumberExpressionNode{
		Value: value,
		Facts: database.NewFacts(span()),
	}, nil
}

func (node *NumberExpressionNode) Visit(visitor *visit.Visitor) {
	visitExpression(visitor, node)

	numberType := &types.NamedTypeNode{
		Name:       "Number",
		Parameters: nil,
		Facts:      database.NewFacts(database.GetSpanFact(node)),
	}
	visitor.Visit(numberType)

	visitor.Constraint(typecheck.GroupConstraint(node, numberType))
}

func (node *NumberExpressionNode) Codegen(c *codegen.Codegen) error {
	span := database.GetSpanFact(node)
	c.WriteString(span, node.Value)
	return nil
}

File: compiler/src/nodes/expressions/number.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenError},
    database::{HiddenNode, Node, NodeRef},
    nodes::{NamedTypeNode, visit_expression},
    syntax::{ParseError, Parser, TokenKind},
    typecheck::GroupConstraint,
    visit::{Visit, Visitor},
};

#[derive(Debug)]
pub struct NumberExpressionNode {
    pub value: String,
}

impl Node for NumberExpressionNode {}

pub fn parse_number_expression(
    parser: &mut Parser<'_>,
) -> Result<NumberExpressionNode, ParseError> {
    let value = parser.token(TokenKind::Number)?;

    Ok(NumberExpressionNode { value })
}

impl Visit for NumberExpressionNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_expression(node, visitor);

        let span = visitor.span(node);
        let number_type = visitor.node(
            span,
            HiddenNode::new(NamedTypeNode {
                name: String::from("Number"),
                parameters: Vec::new(),
            }),
        );

        visitor.visit(&number_type);
        visitor.constraint(GroupConstraint::new(node.clone(), number_type));
    }
}

impl Codegen for NumberExpressionNode {
    fn codegen(&self, ctx: &mut CodegenCtx<'_>) -> Result<(), CodegenError> {
        ctx.write_string(&self.value);
        Ok(())
    }
}
